staging/src/k8s.io/apiserver/pkg/endpoints/filters/constrained_impersonation.go

```diff
@@ -0,0 +1,375 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/server/httplog"
+	"k8s.io/klog/v2"
+	"net/http"
+	"strings"
+)
+
+const (
+	legacyImpersonateVerb        = "impersonate"
+	impersonateUserInfoVerb      = "impersonate:user-info"
+	impersonateSAVerb            = "impersonate:serviceaccount"
+	impersonateNodeVerb          = "impersonate:node"
+	impersonateScheduledNodeVerb = "impersonate:scheduled-node"
+	impersonateOnVerbPrefix      = "impersonate-on:"
+)
+
+// WithConstrainedImpersonation is a filter that will inspect and check requests that attempt to change the user.Info for their requests
+func WithConstrainedImpersonation(handler http.Handler, a authorizer.Authorizer, s runtime.NegotiatedSerializer) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+		requestor, exists := request.UserFrom(ctx)
+		if !exists {
+			responsewriters.InternalError(w, req, errors.New("no user found for request"))
+			return
+		}
+
+		actingAsAttrsList, newUser, err := buildImpersonationAttributes(req.Header, requestor)
+		if err != nil {
+			responsewriters.InternalError(w, req, err)
+			return
+		}
+		if len(actingAsAttrsList) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		impersonateVerb := legacyImpersonateVerb
```

Comment:
prefer placeholder value here and assign `legacyImpersonateVerb` after we actually  use that legacy impersonation

Reply:
updated